### PR TITLE
Start using generic types for states

### PR DESCRIPTION
**Pull request by mfal**
Start using generic types for state values and their respective handler (in a backwards-compatible way).

This makes handling with states more type-safe. 

```typescript
subscribe<boolean>(pattern: "some.id", ({state}) => {state.val.subString(0,1)});
```

The above example will not compile, since `state.val` is now a boolean.

### CODE REVIEW

File: lib/javascript.d.ts [hunk 1]
```diff
@@ -195,10 +200,10 @@ declare global {
 		type Object = StateObject | ChannelObject | DeviceObject | OtherObject;
 		type PartialObject = PartialStateObject | PartialChannelObject | PartialDeviceObject | PartialOtherObject;
 
-		type GetStateCallback = (err: string | null, state?: State) => void;
-		type SetStateCallback = (err: string | null, id?: string) => void;
+		type GetStateCallback<T = any> = (err: string | null, state?: State<T>) => void;
+		type SetStateCallback<T = any> = (err: string | null, id?: string) => void;
```
Comment (AlCalzone) [inline lib/javascript.d.ts]: T is unused. No need for it here.

File: lib/javascript.d.ts [hunk 2]
```diff
@@ -195,10 +200,10 @@ declare global {
 		type Object = StateObject | ChannelObject | DeviceObject | OtherObject;
 		type PartialObject = PartialStateObject | PartialChannelObject | PartialDeviceObject | PartialOtherObject;
 
-		type GetStateCallback = (err: string | null, state?: State) => void;
-		type SetStateCallback = (err: string | null, id?: string) => void;
+		type GetStateCallback<T = any> = (err: string | null, state?: State<T>) => void;
```
Comment (AlCalzone) [inline lib/javascript.d.ts]: I'm torn on this. You can achieve the same by just asserting the type on the callback parameter, e.g.:
```ts
getState("id", (err, state?: iobJS.State<boolean>) => { /* ... */ });
```
This way the generic is more directly associated with the parameter it belongs to.

But I can see the benefit of just writing `getState<boolean>(...)` and not caring about the parameter type names. So let's do this.

File: lib/javascript.d.ts [hunk 3]
```diff
@@ -195,10 +200,10 @@ declare global {
...
+		type SetStateCallback<T = any> = (err: string | null, id?: string) => void;
 
-		type StateChangeHandler = (obj: ChangedStateObject) => void;
+		type StateChangeHandler<T = any> = (obj: ChangedStateObject<T>) => void;
```
Comment (AlCalzone) [inline lib/javascript.d.ts]: This I don't like. You can achieve the same by just asserting the type on the callback parameter, e.g.:
```ts
on("state.id", (obj: iobJS.ChangedStateObject<boolean, string>) => { /* ... */ });
```
This way the generic is more directly associated with the parameter it belongs to. With a generic type on the method itself, it is not immediately clear that it influences a parameter in the callback method.

File: lib/javascript.d.ts [hunk 4]
```diff
@@ -341,17 +346,17 @@ declare global {
...
 			 * Sets all queried states to the given value.
 			 */
-			setState: (id: string, state: string | number | boolean | State | Partial<State>, ack?: boolean, callback?: SetStateCallback) => this;
+			setState: <T extends string | number | boolean>(id: string, state: T | State<T> | Partial<State<T>>, ack?: boolean, callback?: SetStateCallback) => this;
 
 			/**
 			 * Subscribes the given callback to changes of the matched states.
 			 */
-			on: (callback: StateChangeHandler) => this;
+			on: <T = any>(callback: StateChangeHandler<T>) => this;
```
Comment (AlCalzone) [inline lib/javascript.d.ts]: No. See comment on `StateChangeHandler`

File: lib/javascript.d.ts [hunk 5]
```diff
@@ -556,18 +561,18 @@ declare global {
 	/**
 	 * Subscribe to changes of the matched states.
 	 */
-	function on(pattern: string | RegExp | string[], handler: iobJS.StateChangeHandler): any;
-	function on(
+	function on<T = any>(pattern: string | RegExp | string[], handler: iobJS.StateChangeHandler<T>): any;
```
Comment (AlCalzone) [inline lib/javascript.d.ts]: See comment on `StateChangeHandler`.

File: lib/javascript.d.ts [hunk 6]
```diff
@@ -556,18 +561,18 @@ declare global {
...
+	function on<T = any>(
```
Comment (AlCalzone) [inline lib/javascript.d.ts]: See comment on `StateChangeHandler`.

File: lib/javascript.d.ts [hunk 7]
```diff
@@ -556,18 +561,18 @@ declare global {
...
 		astroOrScheduleOrOptions: iobJS.AstroSchedule | iobJS.SubscribeTime | iobJS.SubscribeOptions, 
-		handler: iobJS.StateChangeHandler
+		handler: iobJS.StateChangeHandler<T>
 	): any;
 	/**
 	 * Subscribe to changes of the matched states.
 	 */
-	function subscribe(pattern: string | RegExp | string[], handler: iobJS.StateChangeHandler): any;
-	function subscribe(
+	function subscribe<T = any>(pattern: string | RegExp | string[], handler: iobJS.StateChangeHandler<T>): any;
```
Comment (AlCalzone) [inline lib/javascript.d.ts]: See comment on `StateChangeHandler`.

File: lib/javascript.d.ts [hunk 8]
```diff
@@ -556,18 +561,18 @@ declare global {
...
+	function subscribe<T = any>(
```
Comment (AlCalzone) [inline lib/javascript.d.ts]: See comment on `StateChangeHandler`.

File: lib/javascript.d.ts (hunk 2 above)
Comment (AlCalzone) [reply lib/javascript.d.ts]: Regarding #248 I think the callback parameter should be `State<T> | AbsentState`, shouldn't it?

File: lib/javascript.d.ts [hunk 9]
```diff
@@ -341,17 +346,17 @@ declare global {
 			 * this can be called synchronously and immediately returns the state.
 			 * Otherwise you need to provide a callback.
 			 */
-			getState: (callback?: GetStateCallback) => void | State;
+			getState: <T = any>(callback?: GetStateCallback<T>) => void | State<T> | AbsentState;
```
Comment (AlCalzone) [inline lib/javascript.d.ts]: No. See comment in #248

File: lib/javascript.d.ts [hunk 10]
```diff
@@ -341,17 +346,17 @@ declare global {
...
 
 			/**
 			 * Sets all queried states to the given value.
 			 */
-			setState: (id: string, state: string | number | boolean | State | Partial<State>, ack?: boolean, callback?: SetStateCallback) => this;
+			setState: <T extends string | number | boolean>(id: string, state: T | State<T> | Partial<State<T>>, ack?: boolean, callback?: SetStateCallback) => this;
```
Comment (AlCalzone) [inline lib/javascript.d.ts]: I guess this is more elegant this way, so ok.

Comment (AlCalzone) [reply lib/javascript.d.ts]: But let's add `Record<string, any>` for object-type states and `(string | number | boolean)[]` for array-type states.
We should probably extract that constraint into its own type, e.g. `StateValue` and use it to constrain the Parameter in `State` and `ChangedStateObject`.

File: lib/javascript.d.ts [hunk 11]
```diff
@@ -643,8 +648,8 @@ declare global {
 	 * this can be called synchronously and immediately returns the state.
 	 * Otherwise you need to provide a callback.
 	 */
-	function getState(id: string, callback: iobJS.GetStateCallback): void;
-	function getState(id: string): iobJS.State;
+	function getState<T = any>(id: string, callback: iobJS.GetStateCallback<T>): void;
```
Comment (AlCalzone) [inline lib/javascript.d.ts]: I think T should ahve the same constraint as in getState.

File: lib/javascript.d.ts [hunk 12]
```diff
@@ -23,9 +23,9 @@ declare global {
 			sensor_reports_error = 0x84,
 		}
 
-		interface State {
+		interface State<T = any> {
```
[the remark this answers is not in the document]
Comment (mfal) [reply lib/javascript.d.ts]: So stick to `any`?

[the remark this answers is not in the document]
Comment (AlCalzone) [reply lib/javascript.d.ts]: Yes, stick to `any` for now. `unknown` needs at least TS 2.9, which does not seem to be included with the editor.

**Comment by mfal** [comment]
Updated!

**Comment by AlCalzone** [comment]
I added a few more changes where `any` was still used. Also, `null` is a valid state value.

**Pull request merged**